### DISCUSSION TITLE
How to effectively work with range diffs with FormDiff being retired?

### DISCUSSION THREAD

**Discussion by crypto-rsa** [gitextensions/gitextensions#10081] [Q&A]
One of the recent great additions to GE was `range-diff`, shown in `FormDiff`. It's a real time-saver for me. In our company, we do iterative code reviews, where the reviewer suggests changes and the PR author applies them and then _rebases_ and pushes the new branch. So after I finish a code review, I reset a local branch to the remote one. Once the PR author pushes, I can then compare the local branch to the remote one - and the `range-diff` nicely shows all the changes, including the order of commits and changes in commit messages.
So I was very [disappointed](https://github.com/gitextensions/gitextensions/pull/10072#issuecomment-1189894001) to learn that it has been removed from `FormDiff` in #10072. I know it can be displayed directly in the revision grid when I select two revisions, but
1) I can't browse any other commit while the diff is displayed
2) (more importantly) I can't use the _Select as BASE to compare_ command to "remember" the first revision and then search the revision grid for the second one. I need to select it right after the first one which effectively forbids the use of quick search or other navigational commands. The only way to find the second one is manual scrolling.

To me, these are detrimental changes to an otherwise awesome piece of software. Can anyone explain the reasoning behind them, or suggest an alternative workflow?

**Comment by gerhardol** [comment]
I use range-diff in the same way.

You can open a new instance and view the diff there.
This is the way I want to go to minimize duplicate implementation and maintenance. #10072 was fixing an unoticed bug and the BASE handling was confusing before.

Even if I would like to obsolete FormDiff, this could be added in a correct way, by handling the "selected" BASE correctly.
* Supply all three revisions (A, BASE, B) to FileStatusList as well as HEAD
* FormDiff to use similar interface as RevDiff, allow multi diff
* If three revisions are selected make a special handling for three selected, check if middle is a a parent for both A and B (not checking if base for A and B is same as BASE).

**Reply by crypto-rsa** [reply]
Using a second instance is obviously a solution for problem 1 above (though not very convenient) but it doesn't help at all with problem 2. How can I see a `range-diff` of two revisions which are many screens apart in the revision grid?

**Reply by gerhardol** [reply]
Ctrl click, you can do that in side panel too.

**Reply by crypto-rsa** [reply]
OK, thanks for your answers. The new behaviour and the suggested workarounds are still very much inferior to the original behaviour so for now, I'm sticking with the latest working version.

**Reply by gerhardol** [reply]
You may submit a PR to change the implementation too...

**Reply by crypto-rsa** [reply]
I might take a look at it at some point but I am not sure what's the intended way forward now that `range-diff` has been removed from `FormDiff`.

**Reply by gerhardol** [reply]
> I might take a look at it at some point but I am not sure what's the intended way forward now that `range-diff` has been removed from `FormDiff`.

See my response above. Use `FileStatusList:SetDiffsAsync()` from FormDiff (or a a new SetDiffs with HEAD).
